apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.SessionPoolKey.cs

```diff
@@ -0,0 +1,88 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Cloud.Spanner.Common.V1;
+using System;
+
+namespace Google.Cloud.Spanner.V1;
+public sealed partial class SessionPool
+{
+    /// <summary>
+    /// Aspects under which the sessions are pooled.
+    /// </summary>
+    public class SessionPoolKey : IEquatable<SessionPoolKey>
+    {
+        /// <summary>
+        /// The database for which sessions are pooled on the associated session pool.
+        /// </summary>
+        public DatabaseName DatabaseName { get; }
+
+        /// <summary>
+        /// The database role of the sessions pooled on the associated session pool.
+        /// </summary>
+        public string DatabaseRole { get; }
+
+        /// <summary>
+        ///  Instantiates a new instance of the <see cref="SessionPoolKey"/> object.
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
+        /// <param name="databaseRole">The database role for the associated pool. May be null
+        /// in which case the default (public) database role is used if fine grained access control
+        /// is enabled; otherwise the database role is ignored.</param>
+        private SessionPoolKey(DatabaseName databaseName, string databaseRole)
+        {
+            DatabaseName = GaxPreconditions.CheckNotNull(databaseName, nameof(databaseName));
+            DatabaseRole = databaseRole;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database name. 
+        /// </summary>
+        /// <param name="databaseName">The database name for the associated pool. Must not be null</param>
+        /// <returns></returns>
+        public static SessionPoolKey Create(DatabaseName databaseName) =>
+            new SessionPoolKey(databaseName, null);
+
+        /// <summary>
+        /// Creates a <see cref="SessionPoolKey"/> with the given database role. 
+        /// </summary>
+        /// <param name="databaseRole">The database role for the associated pool. May be null
+        /// in which case the default (public) database role is used if fine grained access control
+        /// is enabled; otherwise the database role is ignored.</param>
+        /// <returns></returns>
+        public SessionPoolKey WithDatabaseRole(string databaseRole) =>
+            new SessionPoolKey(DatabaseName, databaseRole);
+
+        /// <inheritdoc/>
+        public override string ToString() =>
+            $"DatabaseName: {DatabaseName}; DatabaseRole: {DatabaseRole}";
+
+        /// <inheritdoc/>
+        public override int GetHashCode() =>
+            DatabaseRole is null ? DatabaseName.GetHashCode() :
```

Comment:
This can be a single line with the null conditional operator:

```csharp
GaxEqualityHelpers.CombineHashCodes(DatabaseName.GetHashCode(), DatabaseRole?.GetHashCode());
```

Note that I've reordered them so that we use name before role everywhere.